diff --git a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_driver.c b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_driver.c
index ebd316e..9c11184 100644
--- a/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_driver.c
+++ b/drivers/mxc/gpu-viv/hal/os/linux/kernel/gc_hal_kernel_driver.c
@@ -1343,6 +1343,7 @@ static int gpu_runtime_resume(struct device *dev)
 	return 0;
 }
 
+#ifdef CONFIG_PM_SLEEP
 static int gpu_system_suspend(struct device *dev)
 {
 	pm_message_t state={0};
@@ -1353,6 +1354,7 @@ static int gpu_system_resume(struct device *dev)
 {
 	return gpu_resume(to_platform_device(dev));
 }
+#endif
 
 static const struct dev_pm_ops gpu_pm_ops = {
 	SET_RUNTIME_PM_OPS(gpu_runtime_suspend, gpu_runtime_resume, NULL)
diff --git a/drivers/mxc/ipu3/ipu_common.c b/drivers/mxc/ipu3/ipu_common.c
index 1ae9b7f..0cd49a6 100644
--- a/drivers/mxc/ipu3/ipu_common.c
+++ b/drivers/mxc/ipu3/ipu_common.c
@@ -3049,6 +3049,7 @@ bool ipu_ch_param_bad_alpha_pos(uint32_t pixel_fmt)
 EXPORT_SYMBOL(ipu_ch_param_bad_alpha_pos);
 
 #ifdef CONFIG_PM
+#ifdef CONFIG_PM_SLEEP
 static int ipu_suspend(struct device *dev)
 {
 	struct ipu_soc *ipu = dev_get_drvdata(dev);
@@ -3077,6 +3078,7 @@ static int ipu_resume(struct device *dev)
 	dev_dbg(dev, "ipu resume.\n");
 	return 0;
 }
+#endif
 
 int ipu_runtime_suspend(struct device *dev)
 {
diff --git a/drivers/usb/phy/phy-mxs-usb.c b/drivers/usb/phy/phy-mxs-usb.c
index 0423d56..e96947f 100644
--- a/drivers/usb/phy/phy-mxs-usb.c
+++ b/drivers/usb/phy/phy-mxs-usb.c
@@ -539,6 +539,7 @@ static int mxs_phy_remove(struct platform_device *pdev)
 	return 0;
 }
 
+#ifdef CONFIG_PM_SLEEP
 static int mxs_phy_system_suspend(struct device *dev)
 {
 	struct mxs_phy *mxs_phy = dev_get_drvdata(dev);
@@ -558,6 +559,7 @@ static int mxs_phy_system_resume(struct device *dev)
 
 	return 0;
 }
+#endif
 
 SIMPLE_DEV_PM_OPS(mxs_phy_pm, mxs_phy_system_suspend, mxs_phy_system_resume);
 
