 arch/arm/plat-mxc/clock.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/arm/plat-mxc/clock.c b/arch/arm/plat-mxc/clock.c
index 93347eb..1aa2664 100755
--- a/arch/arm/plat-mxc/clock.c
+++ b/arch/arm/plat-mxc/clock.c
@@ -58,12 +58,12 @@ static void __clk_disable(struct clk *clk)
 {
 	if (clk == NULL || IS_ERR(clk))
 		return;
-
+#ifdef CONFIG_CLK_DEBUG
 	if (!clk->usecount) {
 		WARN(1, "clock enable/disable mismatch! clk  %s\n", clk->name);
 		return;
 	}
-
+#endif
 	if (!(--clk->usecount)) {
 		if (clk->disable)
 			clk->disable(clk);
-- 
1.8.4.rc3


