diff --git a/include/configs/udoo.h b/include/configs/udoo.h
index 7632bb7..73d6e84 100644
--- a/include/configs/udoo.h
+++ b/include/configs/udoo.h
@@ -70,7 +70,7 @@
 #define CONFIG_BAUDRATE                        115200
 
 /* HDMI or LVDS Display Configuration */
-#define CONFIG_VIDEO_ENABLED
+#undef CONFIG_VIDEO_ENABLED
 #ifdef CONFIG_VIDEO_ENABLED
 #define CONFIG_VIDEO
 #define CONFIG_VIDEO_IPUV
