diff -ura git.orig/psplash-colors.h git/psplash-colors.h
--- git.orig/psplash-colors.h	2014-08-24 14:05:20.648242439 +0200
+++ git/psplash-colors.h	2014-08-24 14:05:53.504060646 +0200
@@ -20,10 +20,10 @@
 #define _HAVE_PSPLASH_COLORS_H
 
 /* This is the overall background color */
-#define PSPLASH_BACKGROUND_COLOR 0xec,0xec,0xe1
+#define PSPLASH_BACKGROUND_COLOR 0x00,0x00,0x00
 
 /* This is the color of any text output */
-#define PSPLASH_TEXT_COLOR 0x6d,0x6d,0x70
+#define PSPLASH_TEXT_COLOR 0xff,0xff,0xff
 
 /* This is the color of the progress bar indicator */
 #define PSPLASH_BAR_COLOR 0x6d,0x6d,0x70
