diff --git a/psplash.c b/psplash.c
index 09cf0d0..f8bde20 100644
--- a/psplash.c
+++ b/psplash.c
@@ -269,8 +269,8 @@ main (int argc, char** argv)
 
   /* Draw the Poky logo  */
   psplash_fb_draw_image (fb, 
-			 (fb->width  - POKY_IMG_WIDTH)/2, 
-			 ((fb->height * 5) / 6 - POKY_IMG_HEIGHT)/2,
+			 (fb->width  - POKY_IMG_WIDTH)/2,
+			 (fb->height - POKY_IMG_HEIGHT)/2,
 			 POKY_IMG_WIDTH,
 			 POKY_IMG_HEIGHT,
 			 POKY_IMG_BYTES_PER_PIXEL,
@@ -287,8 +287,6 @@ main (int argc, char** argv)
 
   psplash_draw_progress (fb, 0);
 
-  psplash_draw_msg (fb, MSG);
-
   psplash_main (fb, pipe_fd, 0);
 
 
