diff -ur nzbget-14.0-testing.orig/Makefile.in nzbget-14.0-testing/Makefile.in
--- nzbget-14.0-testing.orig/Makefile.in        2014-08-29 18:49:37.893096794 +0200
+++ nzbget-14.0-testing/Makefile.in     2014-08-29 18:48:52.393307391 +0200
@@ -630,8 +630,7 @@
        README \
        ChangeLog \
        COPYING \
-       lib/par2/AUTHOR \
-       lib/par2/README
+       lib/par2/AUTHORS
 
 exampleconf_FILES = \
        nzbget.conf
